### PR TITLE
Support loginOAuthProviderWithRedirect flow

### PR DESCRIPTION
**Pull request by carmenlau**
connect #305 

After #334

### CODE REVIEW

File: packages/skygear-sso/lib/oauth.js [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+      this._oauthWindowObserver.subscribe(newWindow),
+      this._oauthResultObserver.subscribe()
+    ]);
+  }).then((result) => {
+    onLoginCompleted();
+    return this.container.auth._authResolve(result);
+  }, (error) => {
+    onLoginCompleted();
+    return Promise.reject(error);
+  });
```
Comment (rickmak) [inline packages/skygear-sso/lib/oauth.js]: Use `finally` instead of calling twice `onLoginCompleted`

File: packages/skygear-sso/lib/oauth.js [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+        _postSSOResultToWindow(window.opener, authorizedUrls);
+        resolve();
+      } else {
+        reject(errorResponseFromMessage('Fail to find opener'));
+      }
+    }, (err) => {
+      reject(err);
+    });
+  });
+}
```
Comment (rickmak) [inline packages/skygear-sso/lib/oauth.js]:
```
export function oauthHandler() {
  return this.container.lambda('sso/config')
    .then((data) => {
      let authorizedUrls = data.authorized_urls; //eslint-disable-line
      if (window.opener) {
        // popup
        _postSSOResultToWindow(window.opener, authorizedUrls);
        return Promise.resolve();
      } else {
        return Promise.reject(errorResponseFromMessage('Fail to find opener'));
      }
    }
  );
}
```

File: packages/skygear-sso/lib/observer.js [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+    this.onMessageReceived = null;
+    this.result = null;
+  }
+
+  subscribe() {
+    this.unsubscribe();
+    this._clearResult();
+    return new Promise((resolve, reject) => {
+      this.onMessageReceived = this._onMessageReceived
+        .bind(this, resolve, reject);
```
Comment (rickmak) [inline packages/skygear-sso/lib/observer.js]: Bind at constructor? and `_onMessageReceived` should return a promise itself.

File: packages/skygear-sso/lib/oauth.js [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+  };
+
+  return this.container.store.getItem('skygear-oauth-is-login')
+    .then((isRedirectFlowCalled) => {
+      if (!isRedirectFlowCalled) {
+        return Promise.resolve();
+      }
+      if (isRedirectFlowCalled) {
+        addOAuthIframe();
+        this.container.store.removeItem('skygear-oauth-is-login');
```
Comment (rickmak) [inline packages/skygear-sso/lib/oauth.js]: `removeItem` return Promise, so there you will want to user Promise.all.

File: packages/skygear-sso/lib/oauth.js [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+ * @return {Promise} promise
+ *
+ * @example
+ * skygear.auth.loginOAuthProviderWithPopup('google').then(...);
+ */
+export function loginOAuthProviderWithPopup(provider, options) {
+  var newWindow = window.open('', '_blank', 'height=700,width=500');
+  this._oauthWindowObserver = this._oauthWindowObserver ||
+    new NewWindowObserver();
+  this._oauthResultObserver = this._oauthResultObserver ||
```
Comment (rickmak) [inline packages/skygear-sso/lib/oauth.js]: I think it is impossible for `_oauthResultObserver` having a non-null value here?

File: packages/skygear-sso/lib/oauth.js [hunk 6]
```diff
@@ -0,0 +1,231 @@
...
+
+  return this.container.store.getItem('skygear-oauth-is-login')
+    .then((isRedirectFlowCalled) => {
+      if (!isRedirectFlowCalled) {
+        return Promise.resolve();
+      }
+      if (isRedirectFlowCalled) {
+        addOAuthIframe();
+        this.container.store.removeItem('skygear-oauth-is-login');
+        return this._oauthResultObserver.subscribe();
```
Comment (rickmak) [inline packages/skygear-sso/lib/oauth.js]: We should observe the change before adding the iframe.

**Comment by carmenlau** [comment]
Updated! The other PRs are also rebased. Thanks @rickmak

**Pull request merged**